cypress-io/cypress
# feat: cleanup binary with respect to v8 snapshot dependencies

- Closes #23123

### User facing changelog


n/a

### Additional details


This PR's main goal is to decrease the binary size by removing node module files that are no longer needed since they are included in the V8 snaphshot. Care needs to be taken as not everything in Cypress is executed in a context that can utilize the V8 snapshot. Specifically the places that can't are defined by these entry points:

* The initial code that is used to boot strap the snapshot
* The child process code that is used to load consumer config
* The rewriter worker thread

We determine the full tree using the entry points above using `esbuild`. Care needs to be taken to flag any `require.resolve` dependencies as new entry points since they do not get pulled into the tree by `esbuild`. We flag these using the warnings provided by `esbuild` and then continue to run `esbuild` until we have no more entry points to add. In this way, we can determine all of the files that absolutely need to remain.

So our general thought is to use the files in the snapshot cache as a starting point of what can be removed, filter out the dependencies determined using the above process, and then remove the resulting set of files.

This PR:

* Implements the above algorithm
* Turns back on minify in production for the snapshot JS file. It was originally determined that this doesn't affect performance all that much. While this is true in terms of startup, it *does* affect the size of the snapshot binary file itself, thus reducing the size of the overall binary.
* Perform some additional cleanup to the binary (e.g. intermediary build files, ESM distributions, docs, tests, etc.). The goal here was to get the size of the binary on disk to be close to where it was in previous releases. The zip file size will decrease.
* Clean up any empty directories

### Steps to test


You can use the binaries attached to the last commit to just play around and test this to ensure that nothing is missing that shouldn't be.

### How has the user experience changed?


n/a

### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
